pkg/bloomgateway/bloomgateway.go

```diff
@@ -355,19 +312,250 @@ func (g *Gateway) FilterChunkRefs(ctx context.Context, req *logproto.FilterChunk
+	w.queue.RegisterConsumerConnection(w.id)
+	return nil
+}
+
+func (w *worker) running(ctx context.Context) error {
+	idx := queue.StartIndexWithLocalQueue
+
+	requests := make([]v1.Request, 0, 128)
+	fingerprints := make([]uint64, 0, 1024)
+
+	for ctx.Err() == nil {
+		taskCtx := context.Background()
+		dequeueStart := time.Now()
+		items, newIdx, err := w.queue.DequeueMany(taskCtx, idx, w.id, w.cfg.maxItems, w.cfg.maxWaitTime)
+		w.metrics.dequeueWaitTime.WithLabelValues(w.id).Observe(time.Since(dequeueStart).Seconds())
+		if err != nil {
+			// We only return an error if the queue is stopped and dequeuing did not yield any items
+			if err == queue.ErrStopped && len(items) == 0 {
+				return err
+			}
+			w.metrics.dequeueErrors.WithLabelValues(w.id).Inc()
+			level.Error(w.logger).Log("msg", "failed to dequeue tasks", "err", err, "items", len(items))
+		}
+		idx = newIdx
+
+		if len(items) == 0 {
+			continue
+		}
+		w.metrics.dequeuedTasks.WithLabelValues(w.id).Add(float64(len(items)))
+
+		tasksPerDay := make(map[time.Time][]Task)
+
+		for _, item := range items {
+			task, ok := item.(Task)
+			if !ok {
+				// This really should never happen, because only the bloom gateway itself can enqueue tasks.
+				return errors.Errorf("failed to cast dequeued item to Task: %v", item)
+			}
+			level.Debug(w.logger).Log("msg", "dequeued task", "task", task.ID)
+			w.tasks.Delete(task.ID)
+
+			fromDay, throughDay := task.Bounds()
+
+			if fromDay.Equal(throughDay) {
+				tasksPerDay[fromDay] = append(tasksPerDay[fromDay], task)
+			} else {
+				// split task into separate tasks per day
+				for i := fromDay; i.Before(throughDay); i = i.Add(24 * time.Hour) {
+					r := filterRequestForDay(task.Request, i)
+					t := task.WithRequest(r)
+					tasksPerDay[i] = append(tasksPerDay[i], t)
+				}
+			}
+		}
+
+		for day, tasks := range tasksPerDay {
+			logger := log.With(w.logger, "day", day)
+			level.Debug(logger).Log("msg", "process tasks", "tasks", len(tasks))
+
+			it := newTaskMergeIterator(tasks...)
+
+			fingerprints = fingerprints[:0]
+			for it.Next() {
+				// fingerprints are already sorted. we can skip duplicates by checking
+				// if the next is greater than the previous
+				fp := uint64(it.At().Fp)
+				if len(fingerprints) > 0 && fp <= fingerprints[len(fingerprints)-1] {
+					continue
+				}
+				fingerprints = append(fingerprints, fp)
+			}
+
+			it.Reset()
+
+			storeFetchStart := time.Now()
+			// GetBlockQueriers() waits until all blocks are downloaded and available for querying.
+			// TODO(chaudum): Add API that allows to process blocks as soon as they become available.
+			// This will require to change the taskMergeIterator to a slice of requests so we can seek
+			// to the appropriate fingerprint range within the slice that matches the block's fingerprint range.
+			bqs, err := w.store.GetBlockQueriers(taskCtx, tasks[0].Tenant, day, day.Add(24*time.Hour).Add(-1*time.Nanosecond), fingerprints)
```

Comment:
nit: rename `bqs` for `blockQueriers` for better readability throughout the rest of the code.